Allow a plugin attester for generating a custom report - Fixes #3139

jxyang opened the pull request:
This PR allows a plugin attester to be used for generating a custom report which is embedded in a self-signed certificate (required by attested TLS). Also a plugin verifier can be used to validate such a certificate accordingly.  Fixes #3139 

Signed-off-by: Xuejun Yang <xuejya@microsoft.com>

jxyang commented:
bors try

jxyang commented:
bors try

yentsanglee commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +197,121 @@ oe_result_t oe_verify_attestation_certificate(
+
+    // determine the size of the extension
+    if (oe_cert_find_extension(
+            &cert, (const char*)oid_oe_report, NULL, &report_size) !=
```
New oid needs to be defined for embedded evidence.

jxyang replied:
I think it's fine to multiplex the oid for both quote-based report and evidence-based report. The buffer contains a oe_report_header_t that identifies which type it is. The added test case uses that to correctly verify a certificate based evidences.

jxyang replied:
Thanks @CodeMonkeyLeet for the context. Added another oid X509_OID_FOR_EVIDENCE_EXT in report.h.

yentsanglee replied:
I do not see the change in report.h?

yentsanglee commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +197,121 @@ oe_result_t oe_verify_attestation_certificate(
+ * @retval OE_FAILURE general failure
+ * @retval other appropriate error code
+ */
+oe_result_t oe_verify_plugin_attestation_certificate(
```
Suggest 'oe_verify_attestation_evidence_certificate'. Same in all occurrences.

Reason:
1. Embedded attestation evidence is generated by oe_get_evidence().

jxyang replied:
I think the goal of this API is to allow different plugins to generate their own certificates. Evidence is just the means to achieve a generic representation of proof provided by different plugins.

jxyang replied:
As this will become a new public API to OE, I am open to suggestions of better names.

dthaler replied:
I don't understand this PR enough to comment, is there a design doc or diagram?   E.g., is the certificate part of the evidence?   (It certainly is in CyRes attestation.)   If its a public API without _sgx_ in the name, then it should be applicable to other TEEs too, and I can't tell yet whether this PR only makes sense for the SGX plugin.

CodeMonkeyLeet replied:
I agree with @yentsanglee's reasoning and left a similar comment in the header files before I saw this one; either suggested name seems fine to me, though Yen Lee's is shorter. I did make some other related suggestions there as well (e.g. `get` vs. `generate`)

jxyang replied:
Renamed to `oe_get_attestation_certificate_with_evidence` as @CodeMonkeyLeet suggested. Also moved the declaration to attester.h

yentsanglee commented on common/sgx/tlsverifier.c:
```diff
@@ -56,6 +57,30 @@ static oe_result_t verify_report_user_data(
+    oe_result_t result = OE_FAILURE;
+    for (int i = (int)claims_length - 1; i >= 0; i--)
+    {
+        if (oe_strcmp(claims[i].name, "public_key") == 0)
```
Even though "public key" is used internally, it's a good idea to define it in a header file and with unique naming (e.g. "_oe_internal_public_key") so it does not conflict with other claim names defined in include/openenclave/bits/evidence.h.

jxyang replied:
Thanks for catching that. Will do.

yentsanglee commented on common/sgx/tlsverifier.c:
```diff
@@ -56,6 +57,30 @@ static oe_result_t verify_report_user_data(
+    oe_result_t result = OE_FAILURE;
+    for (int i = (int)claims_length - 1; i >= 0; i--)
+    {
+        if (oe_strcmp(claims[i].name, "public_key") == 0)
```
Similar to the other comment.. the claim input is a clear buffer and output is a specific name (see PR 3168).

shnwc commented:
Is this part of the efforts to convert all attestation samples from oe_get_report() to oe_get_evidence()?

For better coordination of the attestation related efforts by multiple developers, I suggest that we bring our early proposals to the SIG-attestation meeting, before implementation if possible.

jxyang commented:
@shnwc This is a clear gap that I am trying to fill for a customer scenario. I have talked to members of SIG-attestation before the implementation. And I am happy to present to the whole group.

jxyang commented:
bors try

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +197,121 @@ oe_result_t oe_verify_attestation_certificate(
+ * @retval other appropriate error code
+ */
+oe_result_t oe_verify_plugin_attestation_certificate(
+    uint8_t* cert_in_der,
```
Coding style nit: expand "cert" -> "certificate" throughout.  Similarly "buf" -> "buffer"

jxyang replied:
The sister functions such as `oe_verify_attestation_certificate` use `cert` and `buf` throughout. I prefer to keep the terms consistent.

dthaler commented on include/openenclave/bits/evidence.h:
```diff
@@ -78,6 +78,11 @@ extern const char* OE_REQUIRED_CLAIMS[OE_REQUIRED_CLAIMS_COUNT];
+/**
+ * Additional claims related to TLS certificate
+ */
+#define OE_CERTIFICATE_PUBLIC_KEY "public_key"
```
If this is a claim name then say so in the #define name.  Other claim names all start with OE_CLAIM_ 
Also the others are in lines 61-77, not separated by what they're "related to".   Put here, it implies it's neither required nor optional.  Not sure what that would mean or why it's in neither array.

jxyang replied:
This is what OE attestation plugin framework calls `custom claims`. In fact, it's a stationary claim in attested TLS in the way that:

1) the attester generates a public/private key pair
2) the attester generates a report, using the hash of the public key as the report data
3) the attester generates a certificate with the report embedded.

On the relying party side:
1) the verifier validates the certificate is signed properly
2) the verifier extracts the report and validates it is signed by a TEE (SGX in this case)
3) the verifier validates the hash of the public key in the certificate matches the report data

With all 3 steps passed, the public key is used for subsequent TLS communications.

In short, this claim neither required nor optional for remote attestation purpose. But it is required for attested TLS or any certificate based attestation.

dthaler commented on include/openenclave/host_verify.h:
```diff
@@ -83,6 +84,43 @@ oe_result_t oe_verify_attestation_certificate(
+/**
+ * oe_verify_plugin_attestation_certificate
+ *
+ * This function perform a custom validation on the input certificate. This
```
```suggestion
 * This function performs a custom validation on the input certificate. This
```

dthaler commented on include/openenclave/enclave.h:
```diff
@@ -677,6 +711,43 @@ oe_result_t oe_verify_attestation_certificate(
+/**
+ * oe_verify_plugin_attestation_certificate
+ *
+ * This function perform a custom validation on the input certificate. This
```
```suggestion
 * This function performs a custom validation on the input certificate. This
```

dthaler commented on tests/attestation_plugin_cert/README.md:
```diff
@@ -0,0 +1,19 @@
+
+- **Host Side**
+  1. Create an enclave
+  2. Issue an ecall (get_tls_cert) into enclave for getting a self-signed certificate embedded with a plugin attestation report of the enclave
```
The term "report" is SGX specific.  Is this meant to be SGX specific?

For example, in some evidence formats the evidence is already a certificate (and thus is usable with TLS).  It sounds like the "embedded" point might be SGX-specific?

Also, some authentication protocols use JWTs or CWTs or instead of certificates per se.  Would this require we have separate public APIs for certificates, JWTs, CWTs, etc.?

jxyang replied:
I replaced `report` with `evidence` to be TEE agnostic in this file. These two functions `oe_get_attestation_certificate_with_evidence` and `oe_verify_attestation_certificate_with_evidence`  could be applicable to non-SGX TEEs, but the plugins/formats we have so far are SGX specific.

dthaler commented on include/openenclave/enclave.h:
```diff
@@ -636,6 +636,40 @@ oe_result_t oe_generate_attestation_certificate(
+ *
+ * @return OE_OK on success
+ */
+oe_result_t oe_generate_plugin_attestation_certificate(
```
+1 to preferring "get" instead of "generate"

CodeMonkeyLeet replied:
One additional observation, this PR will conflict with #3198, which changes `evidence_format` to `format_id` and also brings in the API change to `oe_verify_evidence` to add the explicit `format_id` parameter.

shnwc replied:
@CodeMonkeyLeet thanks for catching the conflict. A few files in the `include` and `tests` directories are also changed in PR #3198, and likely will have conflicts.

@jxyang since this PR will invoke the new APIs implemented in PR #3198, can you wait till that PR is merged, and rebase to it?

jxyang replied:
@shnwc Sure. Looks like your PR has more than enough approvals. What's stopping it from being merged?

shnwc replied:
PR #3198 is merged now. Changes in public API takes longer go reach agreement.

jxyang replied:
Thanks. I will rebase.

jxyang replied:
Why does #3198 changes `evidence_format` to `format_id`? Should this API keep using `format_id`?

jxyang commented on include/openenclave/bits/evidence.h:
```diff
@@ -78,6 +78,11 @@ extern const char* OE_REQUIRED_CLAIMS[OE_REQUIRED_CLAIMS_COUNT];
+/**
+ * Additional claims related to TLS certificate
+ */
+#define OE_CERTIFICATE_PUBLIC_KEY "public_key"
```
Moved to `openenclave/internal/cert.h`.

jxyang commented on include/openenclave/enclave.h:
```diff
@@ -636,6 +636,40 @@ oe_result_t oe_generate_attestation_certificate(
     uint8_t** output_cert,
     size_t* output_cert_size);
 
+/**
```
Done.

jxyang commented on include/openenclave/enclave.h:
```diff
@@ -636,6 +636,40 @@ oe_result_t oe_generate_attestation_certificate(
+/**
+ * oe_generate_plugin_attestation_certificate.
+ *
+ * This function generates a self-signed x.509 certificate with an embedded
+ * report generated by a plugin attester from the underlying enclave.
+ *
+ * @param[in] format_id format id for the plugin attester
```
Thanks for the suggestion. Updated accordingly.

jxyang commented on include/openenclave/enclave.h:
```diff
@@ -677,6 +711,43 @@ oe_result_t oe_verify_attestation_certificate(
+ * @retval OE_FAILURE general failure
+ * @retval other appropriate error code
+ */
+oe_result_t oe_verify_plugin_attestation_certificate(
```
Moved.

jxyang commented on include/openenclave/enclave.h:
```diff
@@ -677,6 +711,43 @@ oe_result_t oe_verify_attestation_certificate(
+ * @param[in] claims_length length of the claims array
+ * @param[in] arg caller defined context
+ */
+typedef oe_result_t (*oe_claim_verify_callback_t)(
```
Renamed.

jxyang replied:
Renamed to `oe_verify_claims_callback_t`

dthaler replied:
My opinion is that oe_<verb>_<noun> or oe_<verb>_<type>_<noun> is better than oe_<type>_<verb>_<noun>.   (FYI, the guidance for new APIs that ship in Windows is the same.)   oe_verify_claims_callback_t is good.

jxyang commented:
Note the added test is failing because of #3264.

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +198,131 @@ oe_result_t oe_verify_attestation_certificate(
+    void* arg)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_cert_t cert = {0};
```
https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md says:

> DON'T use abbreviations unless they are already well-known terms known by users (e.g., "app", "info"), or are already required for use by developers (e.g, "min", "max", "args"). Examples of bad use would be num_widgets instead of widget_count, and opt_widgets instead of option_widgets or optional_widgets.

So: cert -> certificate, pub -> public, buf -> buffer

jxyang replied:
IMHO, this guideline is arbitrary because each person may have difference list of `well-known terms`. In addition, the other functions in OE use the above mentioned abbreviations extensively.

dthaler replied:
Yes the only terms that should be considered well known are ones explicitly mentioned as exceptions in the coding style guidelines (by default, nothing).   I agree that other functions in OE are bad and should be cleaned up too, but not in this PR.

dthaler commented on enclave/tls_cert.c:
```diff
@@ -169,6 +175,93 @@ oe_result_t oe_generate_attestation_certificate(
+ *
+ * @param[in] evidence_format The format UUID of the evidence to be generated.
+ *
+ * @param[in] subject_name a string contains an X.509 distinguished
```
```suggestion
 * @param[in] subject_name a string containing an X.509 distinguished
```

dthaler commented on enclave/tls_cert.c:
```diff
@@ -169,6 +175,93 @@ oe_result_t oe_generate_attestation_certificate(
+ * @param[in] subject_name a string contains an X.509 distinguished
+ * name (DN) for customizing the generated certificate. This name is also used
+ * as the issuer name because this is a self-signed certificate
+ * See RFC5280 (https://tools.ietf.org/html/rfc5280) specification for details
```
```suggestion
 * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details
```

dthaler commented on include/openenclave/attestation/attester.h:
```diff
@@ -113,6 +113,40 @@ oe_result_t oe_get_evidence(
+ *
+ * @param[in] evidence_format The format UUID of the evidence to be generated.
+ *
+ * @param[in] subject_name a string contains an X.509 distinguished
```
```suggestion
 * @param[in] subject_name a string containing an X.509 distinguished
```

dthaler commented on include/openenclave/attestation/attester.h:
```diff
@@ -113,6 +113,40 @@ oe_result_t oe_get_evidence(
+ * @param[in] subject_name a string contains an X.509 distinguished
+ * name (DN) for customizing the generated certificate. This name is also used
+ * as the issuer name because this is a self-signed certificate
+ * See RFC5280 (https://tools.ietf.org/html/rfc5280) specification for details
```
```suggestion
 * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details
```

dthaler commented on include/openenclave/attestation/attester.h:
```diff
@@ -113,6 +113,40 @@ oe_result_t oe_get_evidence(
     size_t* endorsements_buffer_size);
 
+/**
+ * oe_get_attestation_certificate_with_evidence
```
My opinion is that this is not an "attestation certificate".   An "attestation certificate" would be one like in DICE where the certificate itself is the evidence format.   This is a certificate with evidence embedded in it.   Suggest something like `oe_get_self_signed_certificate_with_evidence`.

dthaler replied:
Xuejun pointed out in IM that this name is consistent with the existing API oe_verify_attestation_certificate which I claim has the same problem, but the consistency means that this PR is ok and can discuss separately.

dthaler commented on include/openenclave/attestation/verifier.h:
```diff
@@ -165,6 +165,44 @@ oe_result_t oe_verify_evidence(
+    void* arg);
+
+/**
+ * oe_verify_attestation_certificate_with_evidence
```
As before, I think this name should change.  e.g., `oe_verify_self_signed_certificate_with_evidence` or something similar.

dthaler commented on include/openenclave/internal/report.h:
```diff
@@ -80,6 +80,13 @@ OE_STATIC_ASSERT(
 #define X509_OID_FOR_QUOTE_STRING "1.2.840.113556.10.1.1"
 
+// ISO(1).ANSI(2).USA(840).Microsoft(113556).ACC(10).Classes(1).Subclass(2)
+#define X509_OID_FOR_EVIDENCE_EXT                            \
```
My understanding is that evidence defines/typedefs go in evidence.h
Whereas report.h is for legacy report defines/typedefs

In addition, I think the define needs "OE" or "OPENENCLAVE" in the name somewhere since you use this to indicate that the evidence has an OE-specific header in it (that holds the UUID).   If standard OIDs are defined in the future for standard evidence formats, this name would be misleading.

dthaler commented on tests/attestation_plugin_cert/enc/enc.cpp:
```diff
@@ -0,0 +1,336 @@
+            }
+        }
+        // Dump an enclave's unique ID, signer ID and Product ID. They are
+        // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves. In a real
```
In my opinion, SGX specific info should only appear in a file where "sgx" is in the filename or path to the file.  Much of the rest of this file is agnostic.  I'd recommend moving this function (and any other sgx-only code in this file) to a separate file with sgx in the filename.

jxyang replied:
Renamed the function name to `sgx_enclave_claims_verifier`.

dthaler commented on tests/attestation_plugin_cert/host/host.cpp:
```diff
@@ -0,0 +1,219 @@
+                goto done;
+            }
+        }
+        // Dump an enclave's unique ID, signer ID and Product ID. They are
+        // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves. In a real
+        // scenario, custom id checking should be done here
+        else if (
+            strcmp(claim->name, OE_CLAIM_SIGNER_ID) == 0 ||
+            strcmp(claim->name, OE_CLAIM_UNIQUE_ID) == 0 ||
+            strcmp(claim->name, OE_CLAIM_PRODUCT_ID) == 0)
```
Same comment here about separating SGX-specific code from TEE_agnostic code, by moving SGX-only code into a separate file with sgx in the filename or directory name.

dthaler commented on tests/attestation_plugin_cert/host/host.cpp:
```diff
@@ -0,0 +1,219 @@
+#ifdef OE_LINK_SGX_DCAP_QL
+
+#ifdef _WIN32
+    /* This is a workaround for running in Visual Studio 2017 Test Explorer
+     * where the environment variables are not correctly propagated to the
+     * test. This is resolved in Visual Studio 2019 */
```
Should this be inside an _MSC_VER test then?

jxyang replied:
Removed the workaround since we moved to VS 2019 already.

dthaler commented on tests/attestation_plugin_cert/host/host.cpp:
```diff
@@ -0,0 +1,219 @@
+#endif
+
+    oe_result_t result;
+    oe_enclave_t* enclave = NULL;
```
```suggestion
    oe_enclave_t* enclave = nullptr;
```
cpp files should use the C++ keyword `nullptr` instead of NULL

jxyang replied:
Agree nullptr is better. Changed.

dthaler commented on tests/attestation_plugin_cert/host/host.cpp:
```diff
@@ -0,0 +1,219 @@
+
+int main(int argc, const char* argv[])
+{
+#ifdef OE_LINK_SGX_DCAP_QL
```
Other PRs remove this define.  See PR #3273 and #3270 so will need to update this as they are merged

jxyang replied:
Thanks for the info. Removed.

dthaler commented:
FYI, the main comment I consider a blocker is the name of the OID.   Everything else is a strong suggestion not a blocker.

jxyang commented:
> FYI, the main comment I consider a blocker is the name of the OID. Everything else is a strong suggestion not a blocker.

I renamed it to `X509_OID_FOR_OE_EVIDENCE_EXT` and moved to `evidence.h`.

jxyang commented:
bors try

jxyang commented:
bors try

yentsanglee commented:
> > FYI, the main comment I consider a blocker is the name of the OID. Everything else is a strong suggestion not a blocker.
> 
> I renamed it to `X509_OID_FOR_OE_EVIDENCE_EXT` and moved to `evidence.h`.

Ideally, the 2 X509 OIDs are to be defined in the same header file. Probably in a small header file such as oe_oid.h.

jxyang commented:
> > > FYI, the main comment I consider a blocker is the name of the OID. Everything else is a strong suggestion not a blocker.
> > 
> > 
> > I renamed it to `X509_OID_FOR_OE_EVIDENCE_EXT` and moved to `evidence.h`.
> 
> Ideally, the 2 X509 OIDs are to be defined in the same header file. Probably in a small header file such as oe_oid.h.

I see both suggestions are reasonable.

jxyang commented:
bors try

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -56,6 +59,33 @@ static oe_result_t verify_report_user_data(
 }
 
+// Verify there is a matched claim for the public key */
+static oe_result_t verify_public_key_claim(
```
[development guide](https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md) says

> The single underscore (_ ) is reserved for local definitions (static, file-scope definitions). e.g. static oe_result_t _parse_sgx_report_body(..).


```suggestion
static oe_result_t _verify_public_key_claim(
```

jxyang replied:
Done.

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +202,136 @@ oe_result_t oe_verify_attestation_certificate(
 }
+
+/**
+ * oe_verify_attestation_certificate_with_evidence
```
This is a non-SGX specific function, but is in an SGX-only file (unlike oe_get_attestation_certificate_with_evidence which is already in a non-SGX specific file).  Move this outside the sgx directory for symmetry.

jxyang replied:
Moved it under common (non-sgx specific)

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +202,136 @@ oe_result_t oe_verify_attestation_certificate(
+ * certificate before validating this evidence. An optional
+ * claim_verify_callback could be passed in for a calling client to further
+ * validate the claims of the enclave creating the certificate.
+ * @param[in] cert_in_der a pointer to buffer holding certificate contents
```
Recommend spelling out "cert", "pub", "buf" throughout this file per style guidelines

jxyang replied:
Will keep to be compatible with existing code base.

dthaler commented on common/sgx/tlsverifier.c:
```diff
@@ -172,3 +202,136 @@ oe_result_t oe_verify_attestation_certificate(
+    size_t report_size = 0;
+    oe_report_header_t* header = NULL;
+    uint8_t* pub_key_buf = NULL;
+    size_t pub_key_buf_size = KEY_BUFF_SIZE;
```
"buf" vs "BUFF" is similarly inconsistent in terms of # of F's

jxyang replied:
made it two Fs

dthaler commented on tests/attestation_plugin_cert/tls.edl:
```diff
@@ -0,0 +1,17 @@
+#endif
+
+    trusted {
+        public oe_result_t get_tls_cert_signed_with_ec_key([out] unsigned char** data, [out] size_t* data_size);
+        public oe_result_t get_tls_cert_signed_with_rsa_key([out] unsigned char** data, [out] size_t* data_size);
```
```suggestion
        public oe_result_t get_tls_certificate_signed_with_ec_key([out] unsigned char** data, [out] size_t* data_size);
        public oe_result_t get_tls_certificate_signed_with_rsa_key([out] unsigned char** data, [out] size_t* data_size);
```

jxyang replied:
Will keep to be compatible with existing code base.

jxyang commented:
bors try

jxyang commented:
bors try

jxyang commented:
bors try-

jxyang commented:
bors try

CodeMonkeyLeet commented on common/attest_plugin.c:
```diff
@@ -273,6 +306,119 @@ oe_result_t oe_verify_evidence(
+
+done:
+    oe_free(pub_key_buff);
+    for (size_t i = 0; i < claims_length; i++)
```
Consider using the `oe_free_claims` method instead

jxyang replied:
Thanks. Done.

jxyang commented:
bors r+

jxyang commented:
bors r-

jxyang commented:
bors r+
